### PR TITLE
Feature/119 select multiple statuses for status filter

### PR DESCRIPTION
**Pull request by senelesithole**
1. New multi-status filter component.
2. Back-end implementation as well
3. Replaced double-quotes with single-quotes for front-end ( " " => ' : 'typescripts).
4. Replaced == with === ( == => === :typescripts).
5. Added semi-colon, end of line (  => ; :typescript).
6.Please tripple check the pom file before merge.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -238,7 +239,7 @@
                 <artifactId>maven-surefire-plugin</artifactId>
                 <version>${maven-surefire-plugin.version}</version>
                 <configuration>
-                    <skipTests>true</skipTests>
+                    <skipTests>${skipTests}</skipTests>
```
Comment (kevinwallimann) [inline pom.xml]: Why this change? The surefire plugin has to be disabled because we are using the scalatest plugin. The surefire plugin is automatically added by the maven default config

File: pom.xml [hunk 2]
```diff
@@ -67,6 +67,7 @@
 
 
         <skip.docker>true</skip.docker>
+        <skipTests>true</skipTests>
```
Comment (kevinwallimann) [inline pom.xml]: I don't think tests should be skipped by default

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -61,9 +61,9 @@ sparkYarnSink.additionalConfs.spark.executor.memory=2g
 
 #Postgresql properties for connection to trigger metastore
 db.driver=org.postgresql.Driver
-db.url=jdbc:postgresql://
-db.user=
-db.password=
+db.url=jdbc:postgresql://localhost:5432/senele
+db.user=postgres
+db.password=pass
```
Comment (kevinwallimann) [inline src/main/resources/application.properties]: Be careful what you commit :)

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala [hunk 4]
```diff
@@ -41,13 +41,23 @@ trait SearchableTableQuery {
...
         query.filter(table => applyDateTimeRangeFilter(attributes, table.fieldMapping)))
+      val withMultiContains = request.getContainsMultipleFilterAttributes.foldLeft(filteredQuery)((query, attributes) =>
+        query.filter((table => applyContainsMultipleFilter(attributes, table.fieldMapping))))
 
-      val length = filteredQuery.length.result
+//      val length = filteredQuery.length.result
+      val length = withMultiContains.length.result
 
-      val result = filteredQuery
+/*     val result = filteredQuery
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala]: Please don't commit commented code

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/search/ContainsMultipleFilterAttributes.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.trigger.models.search
+
+case class ContainsMultipleFilterAttributes (override val field: String,
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/search/ContainsMultipleFilterAttributes.scala]: I think it should not be called `containsMultiple` but `equalsMultiple`. With `containsMultiple` I would expect a request like this:
```
{ 
...
"containsMultipleFilterAttributes":[{"field":"status","values":["Succeed"]}]
...
}
```
to return all the dag runs with status "Succeeded" because "Succeeded" contains "Succeed". However this is (correctly) not the case, an exact match is needed.

Comment (kevinwallimann) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/search/ContainsMultipleFilterAttributes.scala]: I would also remove everything related to the `StringEqualsFilter` because it's not used anywhere and is now just dead code. If we ever need such a filter, it can be built based on the `ContainsMultipleFilter`

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala [hunk 6]
```diff
@@ -77,6 +87,11 @@ trait SearchableTableQuery {
       applyRangeFilter(tableField, attributes.start, attributes.end)
     }
 
+    private def applyContainsMultipleFilter(attributes: ContainsMultipleFilterAttributes, fieldMapping: Map[String, Rep[_]]): Rep[Boolean] = {
+      val tableField = fieldMapping(attributes.field).asInstanceOf[Rep[String]]
+      tableField inSetBind attributes.values
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala:79]: 👍 for using `inSetBind` instead of `inSet`

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala [hunk 7]
```diff
@@ -243,6 +247,38 @@ class SearchableTableQueryTest extends FlatSpec with Matchers with BeforeAndAfte
...
+    val searchRequest = TableSearchRequest(
+      containsMultipleFilterAttributes = Some(Seq(filter)),
+      sort = None,
+      from = 0,
+      size = 50
+    )
+
+    val result = await(db.run(underTest.search(searchRequest)))
+
+    val expected = TestSearchableData.testSearchableEntities.filter(data => data.stringValue3.contains("foo") || data.stringValue3.contains("bar"))
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala]: Test says "values exactly equal" (which is correct), but here you're using `contains`

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala [hunk 8]
```diff
@@ -243,6 +247,38 @@ class SearchableTableQueryTest extends FlatSpec with Matchers with BeforeAndAfte
...
+
+    result.total should be > 0
+    result.total shouldBe expected.size
+    result.items should contain theSameElementsAs expected
+  }
+
+  it should "not find values that do not contain the search string" in {
+    val filter = ContainsMultipleFilterAttributes(field = TestSearchableTableFieldNames.stringField3, values = List("not-matching-string", "not-matching-string2"))
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala]: In this negative test case, I would actually use strings like "fo" or "ar", to test for the fact that it really is an exact match and not just a contains

File: ui/package.json [hunk 9]
```diff
@@ -38,22 +38,22 @@
...
     "karma-jasmine": "~2.0.1",
     "karma-jasmine-html-reporter": "^1.4.2",
     "protractor": "~5.4.3",
-    "ts-node": "~8.3.0",
+    "puppeteer": "~2.1.0",
+    "ts-node": "^8.3.0",
     "tslint": "~5.18.0",
-    "typescript": "~3.7.5",
-    "puppeteer": "~2.1.0"
+    "typescript": "~3.7.5"
```
Comment (kevinwallimann) [inline ui/package.json:57]: Try not to commit changes like this. `ts-node` has changed from `~8.3.0` to `^8.3.0` (tilde to caret). I guess this was unintentional

File: ui/src/app/app-routing.module.ts [hunk 10]
```diff
@@ -15,14 +15,14 @@
...
-import {LogInGuardService} from "./services/guards/logInGuard.service";
+import {HomeComponent} from './components/home/home.component';
+import {LoginComponent} from './components/login/login.component';
+import {WorkflowsComponent} from './components/workflows/workflows.component';
+import {RunsComponent} from './components/runs/runs.component';
+import {AuthGuardService} from './services/guards/authGuard.service';
+import {LogInGuardService} from './services/guards/logInGuard.service';
 import {routeNames} from './constants/routes.constants';
-import {WorkflowsHomeComponent} from "./components/workflows/workflows-home/workflows-home.component";
+import {WorkflowsHomeComponent} from './components/workflows/workflows-home/workflows-home.component';
```
Comment (kevinwallimann) [inline ui/src/app/app-routing.module.ts:25]: Afaik, there is no technical argument for preferring either double-quotes or single quotes. One could make the argument to favor single quotes because it requires only one keystroke one a english keyboard while double quotes take to, but imo, that doesn't matter too much.
Eventually we will have to solve IDE auto-formatting problems by configuring an auto-formatter. For now I won't fuss much about it if your IDE does it automatically.

File: ui/src/app/app.module.ts [hunk 11]
```diff
@@ -60,13 +61,15 @@ import {NumberRangeFilterComponent} from "./components/runs/filters/number-range
...
-    NumberRangeFilterComponent
+    NumberRangeFilterComponent,
+    MultipleStatusFilterComponent
   ],
   imports: [
     BrowserModule,
     AppRoutingModule,
     ClarityModule,
     FormsModule,
+    ReactiveFormsModule,
```
Comment (kevinwallimann) [inline ui/src/app/app.module.ts]: Do you really need the `ReactiveFormsModule`?

File: ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts [hunk 12]
```diff
@@ -1,23 +1,23 @@
 /*
  * Copyright 2018 ABSA Group Limited
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
+ * Licensed under the Apache License, Version 2.0 (the 'License');
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts]: Hmm did your IDE do that? That is very strange.

File: ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts [hunk 13]
```diff
@@ -59,11 +59,11 @@ export class DatetimeRangeFilterComponent implements ClrDatagridFilterInterface<
   }
 
   isActive(): boolean {
-    return !!this.value.from || !!this.value.to
+    return !!this.value.from || !!this.value.to;
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts:62]: If your IDE did that ok. Otherwise, I wouldn't make changes to files unrelated to the issue, because it makes the review more time-consuming

File: ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+  changes = new Subject<any>();
+
+  constructor() {}
+
+  ngAfterViewInit(): void {
+    this.removeFiltersSubject.subscribe(_ => this.onRemoveFilter());
+  }
+
+  toggleStatuses(statusModel) {
+    if (!statusModel.checked) {
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts]: 1. If you really need the `checked` boolean, I would not store it directly on the `StatusModel`, but maybe create a map from `StatusModel` to boolean. `checked` only has a meaning within this component, but now spills over e.g. to `dagInstanceStatuses.constants.ts` where it has no meaning.

2. I think you don't need `checked` since it is equivalent with `selectedValues.indexOf(statusModel.name) != -1`. In fact, I think you should remove `checked` to eliminate any risk of inconsistency

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.scss [hunk 15]
```diff
@@ -1,6 +1,6 @@
 .status-selectable {
   cursor: pointer;
-  opacity: 0.3;
+  opacity: 0.5;
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/status-filter/status-filter.component.scss]: Why changing the status-filter? Also, I think the status-filter should be removed, because it's not used anywhere. If we will ever need it in the future, we can probably base it on the multiple-status-filter

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.spec.ts [hunk 16]
```diff
@@ -1,7 +1,22 @@
+/*
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/status-filter/status-filter.component.spec.ts]: Remove the file

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.ts [hunk 17]
```diff
@@ -1,8 +1,23 @@
+/*
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/status-filter/status-filter.component.ts]: Remove the file

File: ui/src/app/components/runs/runs.component.html [hunk 18]
```diff
@@ -61,22 +61,34 @@
...
+
+         <clr-dg-filter [clrDgFilter]="multiStatusFilter">
+        <app-multiple-status-filter #multiStatusFilter
+                                    [property]="dagRunColumns.STATUS"
+                                    [removeFiltersSubject]="removeFiltersSubject"
+                                    [statuses]="dagInstanceStatuses.getStatuses()">
+        </app-multiple-status-filter>
+      </clr-dg-filter>
+
+      <!-- <clr-dg-filter [clrDgFilter]="statusFilter">
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/runs.component.html]: Don't commit commented code

File: ui/src/app/components/runs/runs.component.ts [hunk 19]
```diff
@@ -85,7 +86,8 @@ export class RunsComponent implements OnDestroy, AfterViewInit {
       stringEqualsFilterAttributes: this.filters.filter(f => f instanceof StringEqualsFilterAttributes),
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/runs.component.ts]: Remove this line

File: ui/src/app/components/runs/runs.component.ts [hunk 20]
```diff
@@ -96,7 +98,7 @@ export class RunsComponent implements OnDestroy, AfterViewInit {
   }
 
   clearSort() {
-    !!this.sort ? this.columns.find(_ => _.field == this.sort.by).sortOrder = 0 : undefined;
+    !!this.sort ? this.columns.find(_ => _.field === this.sort.by).sortOrder = 0 : undefined;
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/runs.component.ts:99]: You are right about this one, but in the future please don't mix in code which is unrelated to the issue. It makes the review very time-consuming and one PR should deal with only one issue. Changing code unrelated to your issue also pose the risk of causing merge-conflicts with other open PRs.
It's better to open an issue for that which will be solved another time.

File: ui/src/app/models/search/stringEqualsFilterAttributes.model.ts [hunk 21]
```diff
@@ -17,5 +17,6 @@ import {FilterAttributes} from './filterAttributes.model';
```
Comment (kevinwallimann) [inline ui/src/app/models/search/stringEqualsFilterAttributes.model.ts:17]: Remove this file

File: ui/src/app/stores/auth/auth.reducers.ts [hunk 22]
```diff
@@ -13,13 +13,13 @@
...
 
-import * as AuthActions from "./auth.actions";
+import * as AuthActions from './auth.actions';
 import {localStorageKeys} from '../../constants/localStorage.constants';
 
 export interface State {
-  username: string,
-  isAuthenticated: boolean,
-  authenticationFailed: boolean
+  username: string;
```
Comment (kevinwallimann) [inline ui/src/app/stores/auth/auth.reducers.ts:20]: What's bad about using commas instead of semicolons here?

**Review by kevinwallimann — CHANGES_REQUESTED** [review]

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala (hunk 8 above)
Comment (senelesithole) [reply src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala]: spot-on, fixed.

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala (hunk 7 above)
Comment (senelesithole) [reply src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala]: I have changed it to == instead of .equals()
Let me know what you think

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala (hunk 6 above)
Comment (senelesithole) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala:79]: Thank you 😊.

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/search/ContainsMultipleFilterAttributes.scala (hunk 5 above)
Comment (senelesithole) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/search/ContainsMultipleFilterAttributes.scala]: equalsMultiple makes more sense, fixed.
I will also do a single 'commit' where I removed **StringEqualsFilter** from the project as a whole.

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala (hunk 4 above)
Comment (senelesithole) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQuery.scala]: removed

File: src/main/resources/application.properties (hunk 3 above)
Comment (senelesithole) [reply src/main/resources/application.properties]: my bad, thanks for pointing it out.

File: pom.xml (hunk 2 above)
Comment (senelesithole) [reply pom.xml]: fixed (reverted) as well

File: pom.xml (hunk 1 above)
Comment (senelesithole) [reply pom.xml]: I made this change from a hardcoded 'true' to using variable alias, anyways reverted back.

File: ui/package.json (hunk 9 above)
Comment (senelesithole) [reply ui/package.json:57]: how did that occur? :(  fixed.

File: ui/src/app/app-routing.module.ts (hunk 10 above)
Comment (senelesithole) [reply ui/src/app/app-routing.module.ts:25]: Thanks, I would like to keep it as single quotes.

File: ui/src/app/app.module.ts (hunk 11 above)
Comment (senelesithole) [reply ui/src/app/app.module.ts]: For this PR no we do not, maybe for workflow screen we might need it. For now, I have removed it.

File: ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts (hunk 12 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts]: Yes, it did. find and replace in the wrong selection. Fixed.

File: ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts (hunk 13 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts:62]: I will start setting up a typescript plugin (if there is one) for IntelliJ instead of the VS Code.

File: ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts (hunk 14 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts]: Used the second suggestion, thanks.

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.scss (hunk 15 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/status-filter/status-filter.component.scss]: Removed

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.spec.ts (hunk 16 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/status-filter/status-filter.component.spec.ts]: File removed.

File: ui/src/app/components/runs/filters/status-filter/status-filter.component.ts (hunk 17 above)
Comment (senelesithole) [reply ui/src/app/components/runs/filters/status-filter/status-filter.component.ts]: File removed.

File: ui/src/app/components/runs/runs.component.html (hunk 18 above)
Comment (senelesithole) [reply ui/src/app/components/runs/runs.component.html]: Removed.

File: ui/src/app/components/runs/runs.component.ts (hunk 19 above)
Comment (senelesithole) [reply ui/src/app/components/runs/runs.component.ts]: File removed.

File: ui/src/app/components/runs/runs.component.ts (hunk 20 above)
Comment (senelesithole) [reply ui/src/app/components/runs/runs.component.ts:99]: noted with thanks :)

File: ui/src/app/models/search/stringEqualsFilterAttributes.model.ts (hunk 21 above)
Comment (senelesithole) [reply ui/src/app/models/search/stringEqualsFilterAttributes.model.ts:17]: File removed.

File: ui/src/app/stores/auth/auth.reducers.ts (hunk 22 above)
Comment (senelesithole) [reply ui/src/app/stores/auth/auth.reducers.ts:20]: IDE: properties should be separated by semicolon, not commas.

File: pom.xml [hunk 23]
```diff
@@ -29,7 +29,7 @@
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
         <scala.compat.version>2.11</scala.compat.version>
-        <spark.version>2.4.1</spark.version>
+        <spark.version>2.4.4</spark.version>
```
Comment (kevinwallimann) [inline pom.xml]: That isn't part of this issue

File: pom.xml [hunk 24]
```diff
@@ -282,54 +277,7 @@
                     <noCache>true</noCache>
                 </configuration>
             </plugin>
-            <plugin>
```
Comment (kevinwallimann) [inline pom.xml:285]: You didn't remove this part on purpose, did you?

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala [hunk 25]
```diff
@@ -243,6 +211,38 @@ class SearchableTableQueryTest extends FlatSpec with Matchers with BeforeAndAfte
...
+      sort = None,
+      from = 0,
+      size = 50
+    )
+
+    val result = await(db.run(underTest.search(searchRequest)))
+
+    result.total shouldBe 0
+  }
+
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala:245]: Perfect.

File: ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts [hunk 26]
```diff
@@ -39,44 +39,41 @@ export class MultipleStatusFilterComponent implements ClrDatagridFilterInterface
...
-      }
+  ngOnDestroy(): void {
+    this.removeFiltersSubject.unsubscribe();
+  }
+
+toggleStatuses(statusModel) {
+  if (this.selectedValues.indexOf(statusModel.name) < 0) {
+    this.selectedValues = this.selectedValues.concat(statusModel.name);
+  } else {
+    if (this.selectedValues.indexOf(statusModel.name) >= 0) {
```
Comment (kevinwallimann) [inline ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts]: Do you need this if condition here? If you enter this else-block isn't `this.selectedValues.indexOf(statusModel.name)` always >= 0?

Comment (kevinwallimann) [reply ui/src/app/components/runs/filters/multiple-status-filter/multiple-status-filter.component.ts]: Also, please auto-format this file

**Review by kevinwallimann — CHANGES_REQUESTED** [review]
Apart from the comments, this PR looks good to me now.
But I'm still feeling uneasy about all the formatting changes in unrelated files. First, they pose a risk for merge-conflicts. Second, let's say we notice some bug related to this PR in the future. It's going to cost a lot of time to filter out the changes that are actually relevant. I'd like to ask you to revert all the formatting changes to unrelated files and make a separate PR for them. I can help you with that if you want.

File: ui/src/app/stores/auth/auth.reducers.ts (hunk 22 above)
Comment (kevinwallimann) [reply ui/src/app/stores/auth/auth.reducers.ts:20]: ok

File: src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala (hunk 7 above)
Comment (kevinwallimann) [reply src/test/scala/za/co/absa/hyperdrive/trigger/models/tables/SearchableTableQueryTest.scala]: Yes `==` is preferred over `.equals` in Scala, because it handles `null` properly. https://stackoverflow.com/questions/7681161/whats-the-difference-between-and-equals-in-scala

File: ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts (hunk 13 above)
Comment (kevinwallimann) [reply ui/src/app/components/runs/filters/datetime-range-filter/datetime-range-filter.component.ts:62]: I've experimented with prettier, tslint and eslint here: 
https://github.com/AbsaOSS/hyperdrive-trigger/commit/d575ed6e308b4e381a78c7e174a53f6392908e99 and 
https://github.com/AbsaOSS/hyperdrive-trigger/commit/39396b59e77fc7cbc7c1e82584a842cbed8f96cc in one of the prototype branches for the new ui. We should definitely tackle this problem. Ideally, it would work both in IntelliJ and VSCode, if you prefer VSCode

**Comment by senelesithole** [comment]
Okay, I'd like to set up some time with you tomorrow to revert back all the formatting changes ( the git way) rather than manually. 

Thanks a lot, Kevin!
